modepy/tools.py

```diff
@@ -415,4 +439,67 @@ def estimate_lebesgue_constant(n, nodes, shape=None, visualize=False):
 # }}}
 
 
+# {{{ tensor product reshaping
+
+@runtime_checkable
+class Reshapeable(Protocol):
+    shape: Tuple[int, ...]
+
+    def reshape(
+            self: "ReshapeableT", *newshape: Tuple[int, ...], order: str
+            ) -> "ReshapeableT":
+        ...
+
+
+ReshapeableT = TypeVar("ReshapeableT", bound=Reshapeable)
+
+
+def reshape_array_for_tensor_product_space(
+        space: TensorProductSpace, ary: ReshapeableT, axis=-1) -> ReshapeableT:
+    """Return a reshaped view of *ary* that exposes the tensor product nature
+    of the space. Axis number *axis* of *ary* must index coefficients
+    corresponding to a tensor-product-structured basis (e.g. modal or nodal
+    coefficients).
+
+    :arg ary: an array with last dimension having a length matching the
+        :attr:`~modepy.FunctionSpace.space_dim` of *space*.
+    :arg result: *ary* reshaped with axis number *axis* replaced by
+        a tuple of dimensions matching the dimensions of the spaces
+        making up the tensor product. Variation of the represented
+        function along a given dimension will be represented by variation
+        of array entries along the corresponding array axis.
+    """
+    if axis < 0:
+        axis += len(ary.shape)
+    if not (0 <= axis < len(ary.shape)):
+        raise ValueError("invalid axis specified")
+    return ary.reshape(
+            (ary.shape[:axis]
+                + tuple([s.space_dim for s in space.bases])
+                + ary.shape[axis+1:]),
+            order="F")
```

Comment:
Does this need a `.copy(order="C")`? Not sure how it's being used, so maybe not.

Reply:
The intended use case is to reshape group arrays (within `DOFArray`) so that tensor product operators can be applied efficiently (i.e. with way fewer flops), as tested by `test_tensor_product_vdm_dim_by_dim`. Given this use case, I don't think the extra copy would be beneficial. For one, a pyopencl `Array` couldn't even perform the operation. And even if it could, it would impose a cost in memory traffic that just isn't necessary in general.